### DISCUSSION TITLE
System Crafters PNG logo or ASCII art on startup screen?

### DISCUSSION THREAD

**Discussion by szageda** [SystemCrafters/crafted-emacs#299] [Ideas]
Hi,

Do you think it would make sense to add the System Crafters logo or the cogwheel only to Crafted Emacs' start-up screen?  Perhaps in the form of ASCII art?

I've been testing/playing around with Dashboard and here's the result (keeping it minimal):
![Screenshot from 2023-07-10 12-08-15](https://github.com/SystemCrafters/crafted-emacs/assets/84247447/a0bf90e5-3eff-4a25-a18d-a369aa13bb8d)

The logo: https://systemcrafters.net/img/sc_logo.png

Regards,
Ory

**Comment by daviwil** [comment]
The wheel by itself would look cool I think!

**Comment by szageda** [comment]
I've resized the full logo and cropped the wheel.

**Full Logo:** Download - https://e.pcloud.link/publink/show?code=XZsgAfZcl3FdvvX69bKVAlWEfJRuf0z90Yy
![crafted-emacs-dashboard-full-logo](https://github.com/SystemCrafters/crafted-emacs/assets/84247447/e91eb031-317a-457d-aae4-082966716e49)

**Wheel Only:** Download - https://e.pcloud.link/publink/show?code=XZDgAfZC2bxdPtH8iJTHDyuS1Q1nh015KOy
![crafted-emacs-dashboard-logo](https://github.com/SystemCrafters/crafted-emacs/assets/84247447/b84e8163-0389-455f-b6e1-b308626d6d16)

**Dashboard config:**
```Elisp
(setq dashboard-banner-logo-title "Welcome to Crafted Emacs"
      dashboard-startup-banner "~/.config/emacs/system-crafters-logo.png"
      dashboard-center-content t
      dashboard-show-shortcuts nil
      dashboard-display-icons-p t
      dashboard-icon-type 'all-the-icons
      dashboard-set-heading-icons t
      dashboard-set-file-icons t
      dashboard-set-init-info t
      dashboard-set-footer nil)
(dashboard-setup-startup-hook)
(setq initial-buffer-choice (lambda () (get-buffer-create "*dashboard*")))
```

Dashboard apparently also allows you to load a custom text file.  I'll see if I can come up with some ASCII art.  And also test it with `crafted-startup-config`.

**Comment by sthesing** [comment]
Hey, @oryszabo,

I like the idea and look forward to see your experiments with `crafted-startup-config`. 

> I'll see if I can come up with some ASCII art.

If you want (and like it), you can use this:
```ascii
          .000000.          
         .0.    .0.         
       .00.      .00.       
 .000cl.            .lc000. 
.0                        0.
0.        .o0000o.        .0
 00     .0'      '0.     00 
  00   .0          0.   00  
  HHHHH HHHHHHHHHHHH HHHHH  
HHHH HHH HHHHHHHHHHHHHH HHHH
 HHHHHH HHHHHHHHH HHHHHHHH  
  HHH HHHH HHHHHHHHH HHHH   
         HHH HHHHHH         
          HHHHH HH          
```
Feel free to adapt as needed.

**Reply by szageda** [reply]
Unfortunately, I've been carried away with real life stuff, so haven't had the time to work on the idea further.

Thanks for the ASCII wheel though! Dashboard apparently doesn't allow ASCII art; not the way I imagined anyway, so I'll somehow work around the problem.

**Comment by sthesing** [comment]
I experimented with adapting the current `crafted-startup-config` to display a logo (without the use of Dashboard).

It took me quite some time to figure out how to center text that has a face applied to it. 
(That's actually seems to be quite the challenge. The default Emacs welcome text on the splash screen isn't centered to the window width, but in relation to 80 chars and then still it needs to be "eyeballed" to account for the extra space the bigger font size needs... Dashboard does similar shenanigans).
But I found `window-max-chars-per-line` which let's you specify a face and get the actual number of characters that fit in a line with that face. So I managed proper centering like in the screen shots.

![screenshot_2023-09-22T00:32:20](https://github.com/SystemCrafters/crafted-emacs/assets/464966/352a41e0-cbbd-41e0-be59-3611eb828354)
![screenshot_2023-09-22T00:35:25](https://github.com/SystemCrafters/crafted-emacs/assets/464966/f8f0a426-ce75-4d7f-929d-e03811be28db)

My questions now are:

1. Do we want this? (I tend to say yes)
2. If so, do we want the logo to be a clickable link, like the standard Emacs Splash screen does it? (I tend to say no)
3. If so, where do we want to link? The Emacs homepage or the system crafters home page?

In the [respective branch](https://github.com/sthesing/crafted-emacs/tree/v2-startup-splash-logo) in my development fork, I introduced another custom variable `crafted-startup-concise-splash` so the user can turn the logo off. It's supposed to make use of the (currently unused) `concise` parameter in most of `crafted-startup`'s functions. I guess that was the original intent for it...

What do you folks think? Shall I put a bit of polishing work into this or do we not actually want this?

**Reply by jeffbowman** [reply]
I like logos, I had made a pass at it once when I originally wrote the startup code, but abandoned it as a project for future me.  I'd like to see this in Crafted Emacs, and I like the approach of only showing it on GUI versions, while terminal versions just get text (not necessarily the text logo, unless that can be achieved reliably regardless of the terminal - which may be a significant challenge).

For question 2/3, I'm not sure where such a link would go that would be meaningful to the user and consistently available - maybe the info page?  I'm okay with a clickable logo if we can identify a sane and reasonable target, otherwise leaving it as just something nice to look at is fine.

**Reply by jvdydev** [reply]
Logos looking good, regarding the questions I basically agree with your notes.

For links there would be some targets (although I've more than once accidentally activated the link on the Emacs logo and did not appreciate that it took me out of what I was doing :laughing:):
- Crafted Emacs GitHub (although not really any useful, it's not *that* hard to find)
- SystemCrafters Website (for the people who are not (yet) part of the community?)
- The Crafted Emacs Info file (although we do have the "View the manual" link, so somewhat redundant)

**Comment by sthesing** [comment]
Tried an ASCII variant for Emacs in the terminal, too. 
![screenshot_2023-09-27T08:27:40](https://github.com/SystemCrafters/crafted-emacs/assets/464966/800354e2-486e-462b-9113-dd396123d7c7)

It seems I can't handle the line spacing from within the Elisp code. So for this to look good, the user would probably need to change their terminal settings.
I think it's probably better not to display a logo when Emacs is in a non-graphical environment.

**Reply by jeffbowman** [reply]
It looks good to me! :-)

But I'm pretty much okay with how the splash screen currently looks with just text and no logo, so if we need/want to keep doing that for terminal instances, that might be the most straight forward approach, although we would have to detect that in code to turn on/off the logo, which might add to the complexity of the solution.

**Reply by jvdydev** [reply]
Looks good, haven't looked at the code yet, but shouldn't be too hard to figure out if we're in a terminal (see `display-graphic-p`).

**Comment by sthesing** [comment]
> It looks good to me! :-)

Well, if you think it's good enough to be on by default for the terminal...
I'll introduce custom variables to turn the display of the logo separately. The code hasn't become too complex as of now.

I'll prepare a PR. At the end of the day, the question of defaults is only a question of the initial value of the respective custom variables.

**Comment by sthesing** [comment]
With #378, this has been resolved.
